File: packages/csharp/src/Anip.Service/StorageRedaction.cs
namespace Anip.Service;

/// <summary>
/// Strips parameters from low-value audit entries before persistence.
/// Implements SPEC section 6.10.
/// Returns a shallow copy -- does not mutate the input.
/// </summary>
public static class StorageRedaction
{
    /// <summary>
    /// Redacts storage entries by stripping parameters from low-value events.
    /// </summary>
    public static Dictionary<string, object?> RedactEntry(Dictionary<string, object?> entry)
    {
        var result = new Dictionary<string, object?>(entry);

        var ec = result.TryGetValue("event_class", out var ecObj) ? ecObj as string : null;

        if (IsLowValueClass(ec))
        {
            result["parameters"] = null;
            result["storage_redacted"] = true;
        }
        else
        {
            result["storage_redacted"] = false;
        }

        return result;
    }

    private static bool IsLowValueClass(string? ec)
    {
        return ec is "low_risk_success" or "malformed_or_spam" or "repeated_low_value_denial";
    }
}

File: packages/java/anip-service/src/main/java/dev/anip/service/StorageRedaction.java
package dev.anip.service;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public final class StorageRedaction {
    private static final Set<String> LOW_VALUE_CLASSES = Set.of(
        "low_risk_success", "malformed_or_spam", "repeated_low_value_denial"
    );

    private StorageRedaction() {}

    public static Map<String, Object> redactEntry(Map<String, Object> entry) {
        var result = new HashMap<>(entry);
        String ec = (String) result.get("event_class");
        if (ec != null && LOW_VALUE_CLASSES.contains(ec)) {
            result.put("parameters", null);
            result.put("storage_redacted", true);
        } else {
            result.put("storage_redacted", false);
        }
        return result;
    }

    public static boolean isLowValue(String eventClass) {
        return eventClass != null && LOW_VALUE_CLASSES.contains(eventClass);
    }
}
